tests/test_components_timepix.py

```diff
@@ -162,6 +167,23 @@ def test_timepix3_pixel_events(mock_sqs_timepix_run):
     np.testing.assert_allclose(orig_toa.to_numpy() - events['toa'].to_numpy(),
                                walk[(events['tot'] // 25) - 1] * 1e6)
 
+    # Test for data.size exceeding buffer size by monkeypatching
+    # KeyData.ndarray() method to create out-of-bounds data.size.
+    def _ndarray(self, *args, **kwargs):
+        res = orig_ndarray(self, *args, **kwargs)
+
+        if self.source == tpx.raw_instrument_src.source and self.key == 'data.size':
+            res[res.argmax()] += tpx.raw_x_key.entry_shape[0]
+
+        return res
+
+    from extra_data import KeyData
+    orig_ndarray = KeyData.ndarray
+    monkeypatch.setattr(KeyData, 'ndarray', _ndarray)
```

Comment:
This feels like we're doing something pretty complex to test something very simple. I don't see any specific problems it's likely to cause, so feel free to carry on, but... I would probably do this with the `mockdata` framework. Maybe we need some more machinery to make it easier to modify a copy of a file for tests? :thinking:

Reply:
Yes, I'm not particularly proud of this hack, but in the end it seemed to be the most stable to *just* change the returned values. I started with `mockdata`, but subtle things were breaking on other operations and index alignment. In the end EXtra components like this heavily rely on sane data and metadata.

It would be neat to be able to override values for a single test. The EXtra-data test framework is great, but that case is tricky to do. I did not want to add another device with bogus data just for the sake of this test.

Reply:
We've got some tests which do this already, with fixtures which make a fresh copy of a regular example file and then modify it:

https://github.com/European-XFEL/EXtra-data/blob/479ea37389d43515a41736aaa9a62d505cd0ed67/extra_data/tests/test_bad_trains.py#L15-L52

Would something like that work? It's a bit clumsy because you have to explicitly put the full dataset path there to modify it, but that's not exactly difficult.

We could simplify this by making the fixtures for test data function-scoped, so each test gets its own copy of the files anyway and can modify them in place. I made them session scoped to start with for performance, but creating a few temp files shouldn't be that slow. :shrug:

Reply:
No, I think that'll work nicely. Thanks for the hint, TIL.

Reply:
What do you think of 03f19ad808c1a1fd7308548b585328f3e58f6ace? I decided to make a new function-scoped fixture around the original module-scoped data that changes the value back and forth. This should avoid having an entire second run in tmp.

Reply:
I'd rather copy and modify than have a fixture that changes the value back afterwards, to tell you the truth - that sounds like premature optimisation. Copying a small file on a tmpfs should be pretty fast.

If we wanted to do this with directories with many files, we could get smart about linking the ones that we don't want to change.

Reply:
Hu, copying a file back-and-forth is indeed more clever, good idea.

Reply:
Like c48ac4d5cf9681416fcd560857700a88336e8fb0?

Reply:
Yup, LGTM, thanks!